### PR TITLE
[Traceable FSDP2][1/n] Support using user-defined object instance method as hook

### PR DESCRIPTION
**Pull request by yf225**
FSDP2 has this pattern of using user-defined object instance method as hook, and it will throw this error under compile:
`torch._dynamo.exc.Unsupported: call_function UserDefinedObjectVariable(_pre_forward) [FSDPManagedNNModuleVariable(), TupleVariable(), ConstDictVariable()] {}`

This PR adds support for it by always allowing to trace into a UserDefinedObjectVariable that's an instance method (i.e. `MethodType`).

Supersedes https://github.com/pytorch/pytorch/pull/123320.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @LucasLLC

### CODE REVIEW

File: torch/_dynamo/variables/user_defined.py [hunk 1]
```diff
@@ -673,6 +673,12 @@ def call_function(
                 return variables.TorchCtxManagerClassVariable(
                     obj.__class__
                 ).call_function(tx, [var], kwargs)
+
+            func_src = AttrSource(self.source, "__func__") if self.source else None
```
Comment (jansel) [inline torch/_dynamo/variables/user_defined.py:677]: I believe source is required for VariableBuilder to work.  We should call `unimplemented()` if `source=None`

**Review by jansel — CHANGES_REQUESTED** [review]

**Review by jansel — APPROVED** [review]

**Comment by thiagocrepaldi** [comment]
@yf225 out of curiosity, will the user-defined object be represented in the graphModule FX graph? I dont think I understood what you mean by "allow to call into a UserDefinedObjectVariable that's an instance method" in this context

**Comment by yf225** [comment]
> @yf225 out of curiosity, will the user-defined object be represented in the graphModule FX graph? I dont think I understood what you mean by "allow to call into a UserDefinedObjectVariable that's an instance method" in this context

The user-defined object here does not show up in the Dynamo-produced FX graph. "allow to call into a UserDefinedObjectVariable" means allowing Dynamo to trace into it (I updated the desc).

**Pull request closed without merging**